Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift

```diff
@@ -0,0 +1,467 @@
+                return .pauseRequestBodyStream
+            } else {
+                // no state change needed
+                return .wait
+            }
+        }
+    }
+
+    mutating func readEventCaught() -> Action {
+        return .read
+    }
+
+    mutating func errorHappened(_ error: Error) -> Action {
+        switch self.state {
+        case .initialized:
+            preconditionFailure("After the state machine has been initialized, start must be called immidiatly. Thus this state is unreachable")
+        case .running:
+            self.state = .failed(error)
+            return .failRequest(error, closeStream: true)
+        case .finished, .failed:
+            preconditionFailure("If the request is finished or failed, we expect the connection state machine to remove the request immidiatly from its state. Thus this state is unreachable.")
+        }
+    }
+
+    mutating func start() -> Action {
+        guard case .initialized = self.state else {
+            preconditionFailure("Invalid state")
+        }
+
+        self.state = .running(.verifyRequest, .initialized)
+        return .verifyRequest
+    }
+
+    mutating func requestVerified(_ head: HTTPRequestHead) -> Action {
+        guard case .running(.verifyRequest, .initialized) = self.state else {
+            preconditionFailure("Invalid state")
+        }
+
+        guard self.isChannelWritable else {
+            preconditionFailure("Unimplemented. Wait with starting the request here!")
+        }
+
+        if let value = head.headers.first(name: "content-length"), let length = Int(value), length > 0 {
+            self.state = .running(.streaming(expectedBodyLength: length, sentBodyBytes: 0, producer: .producing), .initialized)
+            return .sendRequestHead(head, startBody: true, startReadTimeoutTimer: nil)
+        } else if head.headers.contains(name: "transfer-encoding") {
+            self.state = .running(.streaming(expectedBodyLength: nil, sentBodyBytes: 0, producer: .producing), .initialized)
+            return .sendRequestHead(head, startBody: true, startReadTimeoutTimer: nil)
+        } else {
+            self.state = .running(.endSent, .initialized)
+            return .sendRequestHead(head, startBody: false, startReadTimeoutTimer: self.idleReadTimeout)
+        }
+    }
+
+    mutating func requestVerificationFailed(_ error: Error) -> Action {
+        guard case .running(.verifyRequest, .initialized) = self.state else {
+            preconditionFailure("Invalid state")
+        }
+
+        self.state = .failed(error)
+        return .failRequest(error, closeStream: false)
+    }
+
+    mutating func requestStreamPartReceived(_ part: IOData) -> Action {
+        switch self.state {
+        case .initialized:
+            preconditionFailure("Invalid state: \(self.state)")
+
+        case .running(.verifyRequest, _),
+             .running(.endSent, _):
+            preconditionFailure("Invalid state: \(self.state)")
```

Comment:
Can we collapse these with the one above?

Reply:
Also can you add comments on why they're invalid states?

Reply:
+1